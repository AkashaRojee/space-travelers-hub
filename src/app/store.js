import { configureStore } from '@reduxjs/toolkit';
import rocketsReducer from '../features/rockets/rocketSlice';

export default configureStore({
  reducer: {
    rockets: rocketsReducer,
  },
});
