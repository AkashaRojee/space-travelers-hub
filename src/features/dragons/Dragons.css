@import url('https://fonts.googleapis.com/css2?family=Oswald:wght@200;300;400;500;600;700&display=swap');

* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}
body {
  margin-top: 2rem;
  font-family: 'Oswald', sans-serif;
}
img {
  max-width: 100%;
  object-fit: cover;
}
button {
  line-height: 1.15;
  outline: none;
  font: inherit;
}

.txt-center {
  text-align: center;
}

.flex {
  display: flex;
}
.portrait {
  flex-direction: column;
}
.space-between {
  justify-content: space-between;
}
.cross-center {
  align-items: center;
}
.no-stretch {
  align-self: flex-start;
}
.gap-1 {
  gap: 1rem;
}
.gap-2 {
  gap: 2rem;
}
.gap-3 {
  gap: 3rem;
}
.btn {
  padding: 0.3em 0.4em;
  border-radius: 4px;
  outline: none;
  background: transparent;
  letter-spacing: 1px;
  font-weight: 100;
  font-size: 1rem;
  cursor: pointer;
}
.btn--reserved {
  background: none;
  color: rgb(182, 186, 190);
  border: 0.1px solid rgb(182, 186, 190);
}
.btn--un-reserved {
  border: none;
  color: white;
  background-color: rgb(0, 124, 249);
}
.loading {
  position: absolute;
  top: 60%;
  left: 50%;
  transform: translate(-50%, -50%);
  font-size: 4rem;
}
.label {
  font-size: clamp(1rem, 1vw, 1.5rem);

  letter-spacing: 1px;
  font-weight: 100;
  padding: 0.15em 1em;
  border-radius: 10px;
}
.label--reserved {
  background-color: rgb(0, 163, 183);
  color: white;
  display: block;
  width: 80%;
}
.container {
  width: min(80%, 800px);
  margin: 0 auto;
}
.card {
  padding: 2em 0;
  border: 2px solid rgb(0, 124, 249);
  border-radius: 10px;
  _max-height: 220px;
}
.card + .card {
  margin-top: 2rem;
}
.card__img {
  width: 33%;
}
.card__text {
  font-size: clamp(1rem, 1.25vw, 2rem);
  letter-spacing: 1px;
}
.card__heading {
  color: red;
  letter-spacing: 8px;
  font-weight: 300;
  font-size: clamp(2rem, 2.5vw, 4rem);
}
