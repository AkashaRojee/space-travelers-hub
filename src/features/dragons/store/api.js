// ACTION TYPES
const API_REQUESTED = 'api/requested';
const API_CALL_SUCCESS = 'api/success';
const API_CALL_FAIL = 'api/fail';

// ACTION CREATORS

export const requestAPICall = (payload) => ({
  type: API_REQUESTED,
  payload,
});
export const onAPICallSuccess = () => ({
  type: API_CALL_SUCCESS,
});
export const onAPICallFail = () => ({
  type: API_CALL_FAIL,
});
