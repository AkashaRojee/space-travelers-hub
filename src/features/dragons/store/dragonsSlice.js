import * as apiActions from './api';

// ACTION TYPES
const DRAGONS_REQUESTED = 'dragons/request';
const DRAGONS_LOADED = 'dragons/load';
const DRAGON_RESERVED = 'dragon/:id/reserve';
const DRAGON_UNRESERVED = 'dragon/:id/unreserve';

// ACTION CREATORS
export const loadDragons = () => (dispatch, getState) => {
  const {
    dragons: { lastFetched },
  } = getState();
  return lastFetched
    ? getState()
    : dispatch(
        apiActions.requestAPICall({
          url: 'https://api.spacexdata.com/v3/dragons',
          onStart: DRAGONS_REQUESTED,
          onSuccess: DRAGONS_LOADED,
        })
      );
};
export const reserveDragon = (id) => ({
  type: DRAGON_RESERVED,
  payload: { id },
});
export const unReserveDragon = (id) => ({
  type: DRAGON_UNRESERVED,
  payload: { id },
});

// REDUCER
const initialState = {
  list: [],
  isLoading: false,
  lastFetched: null,
};
const reducer = (state = initialState, action) => {
  const { type, payload } = action;
  if (type === DRAGONS_REQUESTED) return { ...state, isLoading: true };
  if (type === DRAGONS_LOADED) {
    return {
      ...state,
      lastFetched: Date.now(),
      isLoading: false,
      list: payload.map(({ id, name, type, flickr_images: [imgSrc] }) => ({
        id,
        name,
        type,
        imgSrc,
        isReserved: false,
      })),
    };
  }
  if (type === DRAGON_RESERVED) {
    return {
      ...state,
      list: state.list.map((dragon) =>
        dragon.id === payload.id ? { ...dragon, isReserved: true } : dragon
      ),
    };
  }
  if (type === DRAGON_UNRESERVED) {
    return {
      ...state,
      list: state.list.map((dragon) =>
        dragon.id === payload.id ? { ...dragon, isReserved: false } : dragon
      ),
    };
  }
  return state;
};

export default reducer;
