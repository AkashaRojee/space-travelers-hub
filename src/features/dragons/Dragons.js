/* eslint-disable */
import React, { useEffect, Suspense } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { loadDragons, reserveDragon, unReserveDragon } from './store/dragonsSlice';

const RDRow = React.lazy(() => import('../../common/components/RDRow/RDRow'));

const Dragons = () => {
  const { list: dragonsList } = useSelector((state) => state.dragons);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(loadDragons());
  }, []);

  return (
    <Suspense fallback={<div>Loading...</div>}>
      <div className="rd-rows">
        {dragonsList.map(({id, title, type, isReserved, imgSrc }) => (
          <RDRow
            key={id}
            id={id}
            title={title}
            description={`Type: ${type}`}
            image={imgSrc}
            reserved={isReserved}
            actions={{ reserve: reserveDragon, cancel: unReserveDragon }}
          />
        ))}
      </div>
    </Suspense>
  );
};

export default Dragons;
