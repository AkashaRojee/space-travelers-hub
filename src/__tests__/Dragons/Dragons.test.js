import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Dragons from '../../features/dragons/Dragons';
import store from '../../app/store';
import { render, screen } from '@testing-library/react';

const DragonsMock = () => (
  <Provider store={store}>
    <Dragons />
  </Provider>
);

describe('Mission component', () => {
  it('should match the snapshot', () => {
    const shot = renderer.create(<DragonsMock />).toJSON();
    expect(shot).toMatchSnapshot();
  });
});
