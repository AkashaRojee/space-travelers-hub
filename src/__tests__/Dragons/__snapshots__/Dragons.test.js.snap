// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission component should match the snapshot 1`] = `
<div
  className="rd-rows"
/>
`;
